services/storage/series_cursor.go

```diff
@@ -0,0 +1,322 @@
+		}
+
+		c.tags.Set(measurementKey, mm)
+		c.row.key = key
+		c.nf = c.fields
+	}
+
+	c.row.field, c.nf = c.nf[0], c.nf[1:]
+	c.filterset["_field"] = c.row.field
+
+	if c.measurementCond != nil && !evalExprBool(c.measurementCond, c.filterset) {
+		goto RETRY
+	}
+
+	c.tags.Set(fieldKey, []byte(c.row.field))
+
+	if c.cond != nil {
+		// TODO(sgc): lazily evaluate valueCond
+		c.row.valueCond = influxql.Reduce(influxql.CloneExpr(c.cond), c.filterset)
+		if isBooleanLiteral(c.row.valueCond) {
+			// we've reduced the expression to "true"
+			c.row.valueCond = nil
+		}
+	}
+
+	c.row.tags = c.tags.Clone()
+
+	return &c.row
+}
+
+func (c *indexSeriesCursor) Err() error {
+	return c.err
+}
+
+type limitSeriesCursor struct {
+	seriesCursor
+	n, o, c uint64
+}
+
+func newLimitSeriesCursor(ctx context.Context, cur seriesCursor, n, o uint64) *limitSeriesCursor {
+	return &limitSeriesCursor{seriesCursor: cur, o: o, n: n}
+}
+
+func (c *limitSeriesCursor) Next() *seriesRow {
+	if c.o > 0 {
+		for i := uint64(0); i < c.o; i++ {
+			if c.seriesCursor.Next() == nil {
+				break
+			}
+		}
+		c.o = 0
+	}
+
+	if c.c >= c.n {
+		return nil
+	}
+	c.c++
+	return c.seriesCursor.Next()
+}
+
+type groupSeriesCursor struct {
+	seriesCursor
+	rows []seriesRow
+	keys [][]byte
+	f    bool
+}
+
+func newGroupSeriesCursor(ctx context.Context, cur seriesCursor, keys []string) *groupSeriesCursor {
+	g := &groupSeriesCursor{seriesCursor: cur}
+
+	g.keys = make([][]byte, 0, len(keys))
+	for _, k := range keys {
+		g.keys = append(g.keys, []byte(k))
+	}
+
+	return g
+}
+
+func (c *groupSeriesCursor) Next() *seriesRow {
+	if !c.f {
+		c.sort()
+	}
+
+	if len(c.rows) > 0 {
+		row := &c.rows[0]
+		c.rows = c.rows[1:]
+		return row
+	}
+
+	return nil
+}
+
+func (c *groupSeriesCursor) sort() {
+	var rows []seriesRow
+	row := c.seriesCursor.Next()
+	for row != nil {
+		rows = append(rows, *row)
+		row = c.seriesCursor.Next()
+	}
+
+	sort.Slice(rows, func(i, j int) bool {
+		for _, k := range c.keys {
+			ik := rows[i].tags.Get(k)
+			jk := rows[j].tags.Get(k)
+			cmp := bytes.Compare(ik, jk)
+			if cmp == 0 {
+				continue
+			}
+			return cmp == -1
+		}
+
+		return false
+	})
+
+	c.rows = rows
+
+	// free early
+	c.seriesCursor.Close()
+	c.f = true
+}
+
+func isBooleanLiteral(expr influxql.Expr) bool {
+	_, ok := expr.(*influxql.BooleanLiteral)
+	return ok
+}
+
+func toFloatIterator(iter query.Iterator) (query.FloatIterator, error) {
+	sitr, ok := iter.(query.FloatIterator)
+	if !ok {
+		return nil, errors.New("expected FloatIterator")
+	}
+
+	return sitr, nil
+}
+
+func extractFields(itr query.FloatIterator) []string {
+	var a []string
+	for {
+		p, err := itr.Next()
+		if err != nil {
+			return nil
+		} else if p == nil {
+			break
+		} else if f, ok := p.Aux[0].(string); ok {
+			a = append(a, f)
+		}
+	}
+
+	if len(a) == 0 {
+		return a
+	}
+
+	sort.Strings(a)
+	i := 1
+	// TODO(sgc): skip whilst a[i] != a[i-1] to avoid unnecessary copying
```

Comment:
Looks like you already implemented this `TODO`?

Reply:
Removed the TODO; its an unnecessary optimization at this point